org.lflang/src/org/lflang/graph/DirectedGraph.java

```diff
@@ -0,0 +1,288 @@
+ PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ ***************/
+
+package org.lflang.graph;
+
+import java.util.*;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+
+import org.lflang.util.CollectionUtil;
+
+/**
+ * Directed graph that maps nodes to its upstream and downstream neighbors.
+ *
+ * @author {Marten Lohstroh <marten@berkeley.edu>}
+ */
+public class DirectedGraph<T> implements Graph<T> {
+
+    // Note that while both those maps are mutable, the sets
+    // they use as values may not be. They should only be
+    // manipulated through CollectionUtil
+
+    // If a node has no neighbors, it is still in the map with an empty set as a value.
+
+    /**
+     * Adjacency map from vertices to their downstream immediate neighbors.
+     */
+    private final Map<T, Set<T>> downstreamAdjacentNodes = new LinkedHashMap<>();
+
+    /**
+     * Adjacency map from vertices to their upstream immediate neighbors.
+     */
+    private final Map<T, Set<T>> upstreamAdjacentNodes = new LinkedHashMap<>();
+
+
+    /**
+     * Mark the graph to have changed so that any cached analysis is refreshed
+     * accordingly.
+     */
+    protected void graphChanged() {
+        // To be overridden by subclasses that perform analysis.
+    }
+
+
+    /**
+     * Return true if this graph has the given node in it.
+     *
+     * @param node The node to look for.
+     */
+    @Override
+    public boolean hasNode(T node) {
+        return nodes().contains(node);
+    }
+
+
+    /**
+     * Return all immediate upstream neighbors of a given node.
+     *
+     * @param node The node to report the immediate upstream neighbors of.
+     */
+    public Set<T> getUpstreamAdjacentNodes(T node) {
+        return Collections.unmodifiableSet(this.upstreamAdjacentNodes.getOrDefault(node, Set.of()));
+    }
+
+
+    /**
+     * Return all immediate downstream neighbors of a given node.
+     *
+     * @param node The node to report the immediate downstream neighbors of.
+     */
+    public Set<T> getDownstreamAdjacentNodes(T node) {
+        return Collections.unmodifiableSet(this.downstreamAdjacentNodes.getOrDefault(node, Set.of()));
+    }
+
+
+    @Override
+    public void addNode(T node) {
+        this.graphChanged();
+        this.upstreamAdjacentNodes.putIfAbsent(node, Set.of());
+        this.downstreamAdjacentNodes.putIfAbsent(node, Set.of());
+    }
+
+
+    @Override
+    public void removeNode(T node) {
+        this.graphChanged();
+        this.upstreamAdjacentNodes.remove(node);
+        this.downstreamAdjacentNodes.remove(node);
+        CollectionUtil.removeFromValues(this.upstreamAdjacentNodes, node);
```

Comment:
This could use a comment.
```suggestion
        // The node also needs to be removed from the sets that represent connections to the node.
        CollectionUtil.removeFromValues(this.upstreamAdjacentNodes, node);
```